beacon_node/beacon_chain/src/migrate.rs

```diff
@@ -49,18 +48,23 @@ pub trait Migrate<S: Store<E>, E: EthSpec>: Send + Sync + 'static {
 
         // Collect hashes from new_finalized_block back to old_finalized_block (inclusive)
         let mut found_block = false; // hack for `take_until`
-        let newly_finalized_blocks: HashMap<SignedBeaconBlockHash, Slot> = HashMap::from_iter(
+        let maybe_newly_finalized_blocks: Result<HashMap<SignedBeaconBlockHash, Slot>, Error> =
             ParentRootBlockIterator::new(&*store, new_finalized_block_hash.into())
-                .take_while(|(block_hash, _)| {
-                    if found_block {
-                        false
-                    } else {
-                        found_block |= *block_hash == old_finalized_block_hash.into();
-                        true
+                .take_while(|result| match result {
+                    Ok((block_hash, _)) => {
+                        if found_block {
+                            false
+                        } else {
+                            found_block |= *block_hash == old_finalized_block_hash.into();
+                            true
+                        }
                     }
+                    Err(_) => true,
```

Comment:
I think this iterator will never terminate if there's a permanent error condition (e.g., someone deleted the database directory).

I think this is a bit of a general issue with iterators that return a `Result`. A potential work-around is to make the iterator stateful and only return an `Ok(Err(_))` once and then `None` forever afterwards.

Reply:
Nope, [`collect()` short-circuits](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=fn%20main()%20%7B%0A%20%20%20%20let%20v%3A%20Result%3C()%2C%20()%3E%20%3D%20vec!%5BErr(())%5D.into_iter().cycle().collect()%3B%0A%20%20%20%20println!(%22%7B%3A%3F%7D%22%2C%20v)%0A%7D).

Reply:
Oh sorry, I didn't read into this enough and assumed that you were collecting into a `HashMap<Result<_, _>, _>` not a `Result<HashMap<_, _>, _>` because of the `maybe_` prefix.

If you wanted to avoid the `maybe_newly_finalized_blocks` variable you could do this:

```rust
        let newly_finalized_blocks =
            ParentRootBlockIterator::new(&*store, new_finalized_block_hash.into())
                .take_while(|result| match result {
                    Ok((block_hash, _)) => {
                        if found_block {
                            false
                        } else {
                            found_block |= *block_hash == old_finalized_block_hash.into();
                            true
                        }
                    }
                    Err(_) => true,
                })
                .map(|result| result.map(|(block_hash, block)| (block_hash.into(), block.slot())))
                .collect::<Result<HashMap<SignedBeaconBlockHash, Slot>, _>>()?;
```

I just saw @michaelsproul had also suggested it. I'll leave mine here since it's slightly more specific (compiler can't infer all the types).